### PR TITLE
Vermont Charitable Contribution Credit

### LINKED ISSUE #3054: Vermont Charitable Contribution Credit
Vermont Charitable Contribution Credit
Ref: https://law.justia.com/codes/vermont/2022/title-32/chapter-151/section-5822/

### PR DESCRIPTION
**Pull request by Kiki-67**
Fixes #3054

### Summary
📝📄✅

This pull request implements the Vermont charitable contributions credit feature, which allows taxpayers to claim a state income tax credit based on their charitable donations. It adds new parameter, variable, and test files to the `policyengine_us` package.

> _`VT` credit added_
> _New files for parameters_
> _Tests check the `formula`_

### Walkthrough
* Bump minor version and add Vermont charitable contributions credit feature to changelog ([link](https://github.com/PolicyEngine/policyengine-us/pull/3056/files?diff=unified&w=0#diff-5885042762833a4aef51405a4036067e4e6aefa7a5091a2987d1d9d1fc2647e8R1-R4))
* Add parameter files for percentage rate and threshold amount of Vermont charitable contributions credit, with descriptions, values, metadata, and references ([link](https://github.com/PolicyEngine/policyengine-us/pull/3056/files?diff=unified&w=0#diff-52a8cecb6da69f8cf199ac1fb8f63471dcf25bacab6ae204be482a371b21a7cdR1-R15), [link](https://github.com/PolicyEngine/policyengine-us/pull/3056/files?diff=unified&w=0#diff-6ca14cb2ecfaf5fa2f90b6defdca4445c9aefd9983c908011d2ce3f195ceab4eR1-R15))
* Add variable file for Vermont charitable contributions credit, with value type, entity, label, unit, definition period, reference, state code, and formula ([link](https://github.com/PolicyEngine/policyengine-us/pull/3056/files?diff=unified&w=0#diff-de600299c12a4f8ef85b00a50ff1d1dc4f10fc732a984ca8dfb7c20bb7f72127R1-R21))
* Add test cases for Vermont charitable contributions credit, with different input and output values for charitable deduction and credit amount ([link](https://github.com/PolicyEngine/policyengine-us/pull/3056/files?diff=unified&w=0#diff-3b5996b5bffd57f3337bebfe771eed3ac27d889a052734d7c9b167c23861fecbR1-R15))

### CODE REVIEW

File: changelog_entry.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+- bump: minor
+  changes:
+    added:
+    - vermont charitable contributions credit
```
Comment (PavelMakarchuk) [inline changelog_entry.yaml:4]:
```suggestion
    - Vermont charitable contributions credit.
```

File: policyengine_us/parameters/gov/states/vt/tax/income/subtractions/threshold.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+description: Vermont allows charitable contributions credit in excess of this threshold for tax units that do not itemize Vermont deductions.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/vt/tax/income/subtractions/threshold.yaml:1]:
```suggestion
description: Vermont allows charitable contributions credit in excess of this contributions threshold.
```

File: policyengine_us/parameters/gov/states/vt/tax/income/subtractions/percentage.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/vt/tax/income/subtractions/percentage.yaml:1]: rename to rate.yaml

File: policyengine_us/parameters/gov/states/vt/tax/income/subtractions/percentage.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+description: Vermont allows subtraction from charitable contributions credit at this rate.
+values:
+  2021-01-01: 0.05
+
+metadata:
+  period: year
+  unit: /1
+  label: Vermont charity contribution credit rate
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/vt/tax/income/subtractions/percentage.yaml:8]:
```suggestion
  label: Vermont charitable contribution credit rate
```

File: policyengine_us/parameters/gov/states/vt/tax/income/subtractions/percentage.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+description: Vermont allows subtraction from charitable contributions credit at this rate.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/vt/tax/income/subtractions/percentage.yaml:1]:
```suggestion
description: Vermont allows a subtraction under the charitable contributions credit at this rate of charitable contributions in excess of the threshold.
```

File: policyengine_us/variables/gov/states/vt/tax/income/subtractions/vt_charitable_contributions_credit.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/vt/tax/income/subtractions/vt_charitable_contributions_credit.py:1]: put in credits folder

File: policyengine_us/parameters/gov/states/vt/tax/income/subtractions/rate.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/vt/tax/income/subtractions/rate.yaml:1]: move to credits/charitable/rate.yaml

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/vt/tax/income/subtractions/rate.yaml:1]: same for the threshold

File: policyengine_us/tests/policy/baseline/gov/states/vt/tax/income/subtractions/vt_charitable_contributions_credit.yaml
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/vt/tax/income/subtractions/vt_charitable_contributions_credit.yaml:1]: move to credits

File: policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+description: Vermont allows a subtraction under the charitable contributions credit at this rate of charitable contributions in excess of the threshold.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate.yaml:1]:
```suggestion
description: Vermont allows a charitable contributions credit of this fraction of charitable contributions in excess of the threshold.
```

File: policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate.yaml:1]: make a non_refundable.yaml file in the credits folder and include this credit there

Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate.yaml:1]: this credit is capped at $1,000

Comment (Kiki-67) [reply policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate.yaml:1]: added a test for the capped amount

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate.yaml:1]: dont we need a parameter file for this condition?

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate.yaml:1]: I see, 5% of 20,000 is 1,000

File: policyengine_us/parameters/gov/states/vt/tax/income/credits/non_refundable.yaml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+description: Vermont non-refundable tax credits.
+values:
+  2021-01-01:
+    - vt_charitable_contribution_credit
+metadata:
+  unit: list
+  label: Vermont non-refundable tax credits
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/vt/tax/income/credits/non_refundable.yaml:7]: add reference

**Review by PavelMakarchuk — APPROVED** [review]

File: policyengine_us/parameters/gov/states/vt/tax/income/credits/non_refundable.yaml (hunk 6 above)
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/vt/tax/income/credits/non_refundable.yaml:7]: add reference

File: policyengine_us/tests/policy/baseline/gov/states/vt/tax/income/credits/vt_charitable_contributions_credit.yaml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+- name: filer has 1,000 charitable contribution
```
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/vt/tax/income/credits/vt_charitable_contributions_credit.yaml:1]:
```suggestion
- name: Filer has 1,000 of charitable contributions 
```

File: policyengine_us/tests/policy/baseline/gov/states/vt/tax/income/credits/vt_charitable_contributions_credit.yaml [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  period: 2021
+  input:
+    charitable_deduction: 1_000
+    state_code: VT
+  output:
+    vt_charitable_contribution_credit: 50
+
+- name: filer has 30,000 charitable contribution, capped at $1,000
```
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/vt/tax/income/credits/vt_charitable_contributions_credit.yaml:9]:
```suggestion
- name: Filer has 30,000 of charitable contributions, capped at $1,000
```

File: policyengine_us/variables/gov/states/vt/tax/income/credits/vt_charitable_contributions_credit.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    unit = USD
+    definition_period = YEAR
+    reference = (
+        "https://law.justia.com/codes/vermont/2022/title-32/chapter-151/section-5822/"
+        "https://tax.vermont.gov/sites/tax/files/documents/IN-111-2022.pdf#page=1"
+    )
+    defined_for = StateCode.VT
+
+    def formula(tax_unit, period, parameters):
+        charity = tax_unit("charitable_deduction", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/vt/tax/income/credits/vt_charitable_contributions_credit.py:17]:
```suggestion
        charitable_deduction = tax_unit("charitable_deduction", period)
```

File: policyengine_us/variables/gov/states/vt/tax/income/credits/vt_charitable_contributions_credit.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+        p = parameters(period).gov.states.vt.tax.income.credits.charitable
+        subtraction_amount = p.rate * charity
+        capped_subtraction_amount = p.rate * p.threshold
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/vt/tax/income/credits/vt_charitable_contributions_credit.py:20]:
```suggestion
        capped_credit_amount = p.rate * p.threshold
```

File: policyengine_us/parameters/gov/states/vt/tax/income/credits/non_refundable.yaml [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+  reference:
+  - title: 2022 Vermont Statutes Title 32 - Taxation and Finance § 5822-(d)-3
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/vt/tax/income/credits/non_refundable.yaml:9]:
```suggestion
  - title: Vermont Statutes Title 32 - Taxation and Finance § 5822-(d)-3
```

File: policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/threshold.yaml [hunk 12]
```diff
@@ -0,0 +1,15 @@
+description: Vermont allows a charitable contributions credit in excess of this contributions threshold.
+values:
+  2021-01-01: 20_000
+
+metadata:
+  period: year
+  unit: currency-USD
+  label: Vermont charitable contribution credit threshold
+  reference:
+  - title: 2022 Vermont Statutes Title 32 - Taxation and Finance § 5822-(d)-3
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/threshold.yaml:10]:
```suggestion
  - title: Vermont Statutes Title 32 - Taxation and Finance § 5822-(d)-3
```

File: policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate.yaml [hunk 13]
```diff
@@ -0,0 +1,15 @@
+description: Vermont allows a charitable contributions credit of this fraction of charitable contributions in excess of the contributions threshold.
+values:
+  2021-01-01: 0.05
+
+metadata:
+  period: year
+  unit: /1
+  label: Vermont charitable contribution credit rate
+  reference:
+  - title: 2022 Vermont Statutes Title 32 - Taxation and Finance § 5822-(d)-3
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate.yaml:10]:
```suggestion
  - title: Vermont Statutes Title 32 - Taxation and Finance § 5822-(d)-3
```

File: policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/threshold.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/threshold.yaml:1]: rename to cap.yaml

File: policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/threshold.yaml [hunk 14]
```diff
@@ -0,0 +1,15 @@
+description: Vermont allows a charitable contributions credit in excess of this contributions threshold.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/threshold.yaml:1]:
```suggestion
description: Vermont caps the charitable contributions credit to this amount.
```

**Review by PavelMakarchuk — APPROVED** [review]

File: policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate.yaml [hunk 15]
```diff
@@ -0,0 +1,15 @@
+description: Vermont allows a charitable contributions credit of this fraction of charitable contributions in excess of the contributions threshold.
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate.yaml:1]: what threshold?

File: policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/cap.yaml [hunk 16]
```diff
@@ -0,0 +1,15 @@
+description: Vermont caps the charitable contributions credit at this amount.
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/cap.yaml:1]: unify this with rate.yaml as a scale parameter. note it caps *contributions* at this amount, not the credit

File: policyengine_us/variables/gov/states/vt/tax/income/credits/vt_charitable_contributions_credit.py [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+    unit = USD
+    definition_period = YEAR
+    reference = (
+        "https://law.justia.com/codes/vermont/2022/title-32/chapter-151/section-5822/"
+        "https://tax.vermont.gov/sites/tax/files/documents/IN-111-2022.pdf#page=1"
+    )
+    defined_for = StateCode.VT
+
+    def formula(tax_unit, period, parameters):
+        charitable_deduction = tax_unit("charitable_deduction", period)
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/vt/tax/income/credits/vt_charitable_contributions_credit.py:17]: it's charitable contributions (cash and non-cash), not the deduction (which e.g. applies AGI limitations)

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate_threshold.yaml [hunk 18]
```diff
@@ -0,0 +1,23 @@
+description: Vermont caps the charitable contributions at this amount.
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate_threshold.yaml:1]:
```suggestion
description: Vermont provides a credit for this fraction of charitable contributions, depending on the amount.
```

File: policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate_threshold.yaml [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+  - threshold:
+      2018-01-01: 20_000
+    rate:
+      2018-01-01: 0
+metadata:
+  threshold_unit: currency_USD
+  rate_unit: /1
+  threshold_period: year
+  type: marginal_rate
+  label: Vermont charitable contribution credit threshold
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate_threshold.yaml:16]:
```suggestion
  label: Vermont charitable contribution credit rate
```

File: policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate_threshold.yaml
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/vt/tax/income/credits/charitable/rate_threshold.yaml:1]: Move this to `credits/charitable.yaml`

File: policyengine_us/variables/gov/states/vt/tax/income/credits/vt_charitable_contributions_credit.py [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+    unit = USD
+    definition_period = YEAR
+    reference = (
+        "https://law.justia.com/codes/vermont/2022/title-32/chapter-151/section-5822/"
+        "https://tax.vermont.gov/sites/tax/files/documents/IN-111-2022.pdf#page=1"
+    )
+    defined_for = StateCode.VT
+
+    def formula(tax_unit, period, parameters):
+        charitable_contribution = tax_unit("charitable_deduction", period)
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/vt/tax/income/credits/vt_charitable_contributions_credit.py:17]: My comment was not to change the Python variable name without changing the logic. If you have a question about a PR comment please ask it.
```suggestion
        charitable_contributions = add(tax_unit, period, ["charitable_cash_donations", "charitable_non_cash_donations"])
```

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

**Review by MaxGhenis — APPROVED** [review]

**Pull request merged**